dht/i_partitioner.hh

```diff
@@ -648,6 +650,47 @@ future<utils::chunked_vector<partition_range>> split_range_to_single_shard(const
 
 std::unique_ptr<dht::i_partitioner> make_partitioner(sstring name);
 
+// Filter owned tokens using token_range_vector.
+// Conforms to FlattenedConsumerFilterV2 so it can be used
+// with flat_mutation_reader_v2::consume_in_thread.
+class tokens_filter {
+    token_range_vector _sorted_owned_ranges;
+    token_range_vector::iterator _current;
+    token_range_vector::iterator _end;
+    std::optional<token> _prev_token;
+public:
+    // sorted_owned_ranges must be deoverlapped and
+    // sorted by ascending token order.
+    tokens_filter(token_range_vector sorted_owned_ranges)
+        : _sorted_owned_ranges(std::move(sorted_owned_ranges))
+        , _current(_sorted_owned_ranges.begin())
+        , _end(_sorted_owned_ranges.end())
+    {}
+
+    bool operator()(const decorated_key& dk) {
+        const auto& token = dk.token();
+        // Assumes that this call operator is called
+        // by a sorted stream, in sorted token order.
+        assert(!_prev_token || dht::tri_compare(token, *_prev_token) >= 0);
+        _prev_token = token;
+        for (; _current != _end; ++_current) {
+            if (_current->start() && dht::tri_compare(token, _current->start()->value()) < 0) {
+                return false;
+            }
+            if (!_current->end() || dht::tri_compare(token, _current->end()->value()) <= 0) {
+                return true;
+            }
+        }
+        return false;
```

Comment:
we have a implementation of this filter in cleanup_compaction::incremental_owned_ranges_checker. we can probably reuse it there.

Reply:
Maybe we can reuse this implementation which is simpler to understand:

```
            // While token T is after a range Rn, advance the iterator.
            // iterator will be stopped at a range which either overlaps with T (if T belongs to node),
            // or at a range which is after T (if T doesn't belong to this node).
            while (_it != _sorted_owned_ranges.end() && _it->after(t, dht::token_comparator())) {
                _it++;
            }

            return _it != _sorted_owned_ranges.end() && _it->contains(t, dht::token_comparator());
```

A simple unit test would also be benefitial, but can come later. This is a critical procedure.

Reply:
ah good point, I'll look into it.

Reply:
Fixed in v2 by refactoring incremental_owned_ranges_checker out of cleanup compaction